src/util/metadata/side_metadata/global.rs

```diff
@@ -97,6 +99,515 @@ impl SideMetadataSpec {
+                }
+            },
+            |_| {
+                #[cfg(feature = "extreme_assertions")]
+                sanity::verify_store(self, data_addr, metadata);
+            },
+        )
+    }
+
+    #[inline(always)]
+    pub fn compare_exchange_atomic<T: MetadataValue>(
+        &self,
+        data_addr: Address,
+        old_metadata: T,
+        new_metadata: T,
+        success_order: Ordering,
+        failure_order: Ordering,
+    ) -> bool {
+        self.side_metadata_access::<T, _, _, _>(
+            data_addr,
+            || {
+                let meta_addr = address_to_meta_address(self, data_addr);
+                let bits_num_log = self.log_num_of_bits;
+                if bits_num_log < 3 {
+                    let lshift = meta_byte_lshift(self, data_addr);
+                    let mask = meta_byte_mask(self) << lshift;
+
+                    let real_old_byte = unsafe { meta_addr.atomic_load::<AtomicU8>(success_order) };
+                    let expected_old_byte =
+                        (real_old_byte & !mask) | ((old_metadata.to_u8().unwrap()) << lshift);
+                    let expected_new_byte =
+                        (expected_old_byte & !mask) | ((new_metadata.to_u8().unwrap()) << lshift);
+
+                    unsafe {
+                        meta_addr
+                            .compare_exchange::<AtomicU8>(
+                                expected_old_byte,
+                                expected_new_byte,
+                                success_order,
+                                failure_order,
+                            )
+                            .is_ok()
+                    }
+                } else {
+                    unsafe {
+                        T::compare_exchange(
+                            meta_addr,
+                            old_metadata,
+                            new_metadata,
+                            success_order,
+                            failure_order,
+                        )
+                    }
+                    .is_ok()
+                }
+            },
+            |_success| {
+                #[cfg(feature = "extreme_assertions")]
+                if _success {
+                    sanity::verify_store(self, data_addr, new_metadata);
+                }
+            },
+        )
+    }
+
+    #[inline(always)]
+    fn fetch_ops_on_bits<F: Fn(u8) -> u8>(
+        &self,
+        data_addr: Address,
+        meta_addr: Address,
+        set_order: Ordering,
+        fetch_order: Ordering,
+        update: F,
+    ) -> u8 {
+        let lshift = meta_byte_lshift(self, data_addr);
+        let mask = meta_byte_mask(self) << lshift;
+        let mut old_val = unsafe { meta_addr.load::<u8>() };
+        let mut new_sub_val = update((old_val & mask) >> lshift) & (mask >> lshift);
+        let mut new_val = (old_val & !mask) | (new_sub_val << lshift);
+
+        while unsafe {
+            meta_addr
```

Comment:
Similar to `atomic_store`, we can use `fetch_update` instead of hand-written cmpxchg loop.

Reply:
I will make the change.